spread/openstack.go

```diff
@@ -0,0 +1,758 @@
+			if len(s.d.Networks) > 0 {
+				return server.Addresses[s.d.Networks[0]][0].Address, nil
+			}
+			// When no network selected, the we use the first ip in addresses
+			for _, net_val := range server.Addresses {
+				if len(net_val[0].Address) > 0 {
+					return net_val[0].Address, nil
+				}
+			}
+		}
+
+		select {
+		case <-retry.C:
+			debugf("Server %s is taking a while to get IP address...", s)
+		case <-timeout:
+			return "", fmt.Errorf("timeout waiting for Openstack server %s IP address", s)
+		case <-ctx.Done():
+			return "", fmt.Errorf("cannot wait for %s IP address: interrupted", s)
+		}
+	}
+}
+
+func (p *openstackProvider) createMachine(ctx context.Context, system *System) (*openstackServer, error) {
+	debugf("Creating new openstack server for %s...", system.Name)
+
+	name := openstackName()
+	flavorName := openstackDefaultFlavor
+	if system.Plan != "" {
+		flavorName = system.Plan
+	}
+	flavor, err := p.findFlavor(flavorName)
+	if err != nil {
+		return nil, err
+	}
+
+	networks := []nova.ServerNetworks{}
+	if len(system.Networks) == 0 {
+		net, err := p.findFirstNetwork()
+		if err != nil {
+			return nil, err
+		}
+		networks = []nova.ServerNetworks{{NetworkId: net.Id}}
+	}
+
+	for _, netName := range system.Networks {
+		net, err := p.findNetwork(netName)
+		if err != nil {
+			return nil, err
+		}
+		networks = append(networks, nova.ServerNetworks{NetworkId: net.Id})
+	}
+
+	image, err := p.findImage(system.Image)
+	if err != nil {
+		return nil, err
+	}
+
+	availabilityZone, err := p.findAvailabilityZone()
+	if err != nil {
+		return nil, err
+	}
+
+	// cloud init script
+	cloudconfig := fmt.Sprintf(openstackCloudInitScript, p.options.Password)
+
+	// tags to the created instance
+	tags := map[string]string{
+		"spread":   "true",
+		"owner":    strings.ToLower(username()),
+		"reuse":    strconv.FormatBool(p.options.Reuse),
+		"password": p.options.Password,
+	}
+
+	opts := nova.RunServerOpts{
+		Name:             name,
+		FlavorId:         flavor.Id,
+		ImageId:          image.Id,
+		AvailabilityZone: availabilityZone.Name,
+		Networks:         networks,
+		Metadata:         tags,
+		UserData:         []byte(cloudconfig),
+	}
+
+	if len(system.Groups) > 0 {
+		sgNames, err := p.findSecurityGroupNames(system.Groups)
+		if err != nil {
+			return nil, err
+		}
+		opts.SecurityGroupNames = sgNames
+	}
+
+	server, err := p.computeClient.RunServer(opts)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create instance: %v", &openstackError{err})
+	}
+
+	s := &openstackServer{
+		p: p,
+		d: openstackServerData{
+			Id:       server.Id,
+			Name:     name,
+			Flavor:   flavor.Name,
+			Networks: system.Networks,
+			Status:   openstackProvisioning,
+			Created:  time.Now(),
+		},
+
+		system: system,
+	}
+
+	err = p.waitProvision(ctx, s)
+	if err == nil {
+		s.address, err = p.address(ctx, s)
+	}
+	if err == nil {
+		err = p.waitServerBoot(ctx, s)
+	}
+	// TODO: do we need to drop cloud-init script?
```

Comment:
@sergiocazzolato do we need to drop cloud-init script?

Reply:
no, we don't because in this script we configure ssh to allow the root connection using password and also it displays the marker that indicates the machine is ready

Reply:
Those should run in the first boot only, right?